rfcs/rfc-0000-discriminators.md

```diff
@@ -0,0 +1,123 @@
+- This should curb any incentive to sell / steal accounts for "rare usernames".
+
+  Certain "rare" discriminators can also be prohibited.
+
+- Allow users who do not want to be easily discoverable to stay hidden.
+
+  Just a 4-digit discriminator provides 10,000 unique users.
+  Expanding on this, we could for example, let users hide discriminators in servers and similar privacy features.
+
+# Guide-level explanation
+
+Revolt would be switching to a new username system made up of three parts:
+
+```
+Hello, World!#123456
+^ your chosen name
+             ^ separator between chosen name and discriminator
+              ^ discriminator
+```
+
+Discriminators for new users are generated randomly on sign up while existing users will
+get a randomly generated discriminators upon the server software being upgraded.
+
+> Revolt treats usernames as case-insensitive so a full username such as "Jason#1234" is
+> effectively the same as "jasON#1234" but only "Jason#1234" will display in the UI.
+
+Users may:
+
+- Change the case of their username at any moment, such as from "Jason" to "jasON".
+
+  This will never affect their discriminator.
+
+- Change their username, but may have their discriminator regenerated if there is a conflict.
+
+  Such as if you are changing from "Jason#1234" to "Phil", but "Phil#1234" is taken.
+
+### User Opinion
+
+Prior discussions on the forums primarily seem to lean in favour of discriminators being
+implemented into Revolt, the internal team has not raised any complaints against it, and
+the primary community on Revolt ("Revolt Lounge") appears to be somewhat split on the
+issue.
+
+As a recent case study, Discord's change to unique usernames only has generated a lot of
+backlash and criticism from the community, it appears to be overwhemingly negative but
+these experiences are mostly coming from long-term users of Discord. The true opinion of
+the average user has not been gauged yet.
+
+### Usability
+
+This may make it more difficult to understand how to add users on Revolt, however this
+shouldn't be an issue as long as it's clearly communicated through the user interface.
+Revolt already allows a wide range of Unicode in usernames so we are definitely not
+making a significant UX hit towards the usability of the friend system.
+
+> There are further considerations listed at the end of this document which must be
+> addressed if this RFC were to be implemented.
+
+# Reference-level explanation
+
+WIP
+
+This is the technical section of the RFC, it should go over in detail:
+
+- Its interaction with other features
+- How this will be implemented
+- Corner or edge cases
+
+This section should reference the examples in the previous section and disect them in more detail.
+
+# Drawbacks
+
+WIP
+
+Why should this not be added.
+
+# Rationale and alternatives
+
+WIP
+
+- Why is this design the best
+- Are there alternative ways to solve this
+- Could this be done with existing features or existing solutions
```

Comment:
Right off the bat, I'd like to mention how 100% unique usernames often end up creating a thing called an "OG market", where especially notable or rare usernames are taken and sold for real money or other valuables. This is a huge thing in the Minecraft community, where short names (and especially names that are shorter than the current minimum character limit) are insanely rare and valuable to many players. This can be largely mitigated by discriminators, allowing more people to have the username they actually want. rather than allowing certain names to be hogged by traders and people who are just using a particular username for internet points.

I will admit, I may be slightly biased in this regard, because I've had a run-in with this particular community while I had a spotlight in the Minecraft community for a short period. It was Not Fun. But I believe it's still something to be wary of.

Reply:
I might also add that discriminators mitigate the issue where people might create accounts to try Revolt, and then:
1. they decide they don't like Revolt
2. they forget their password and don't bother recovering it
3. they just straight up forget they have an account

...which leads to the account being abandoned, leaving the given username unusable essentially forever. Though, that's assuming abandoned Revolt accounts aren't already automatically erased or anything.

Reply:
> they see revolt has this same system already
...but it *doesn't*...? It's unique IDs, not usernames.

That being said, it's worth asking how the current system can be changed to a discriminator based one.

Reply:
> in light of the heavy backlash on discord, this problem has raised in priority and gravity.

yea, that's pretty much why I've been commenting here lol. I'm hoping increased activity like this will help speed up implementation of the system down the line. I'd love nothing more than to see an open-source, privacy-focused platform like Revolt become people's main place to chat and build communities.

Reply:
> ...but it _doesn't_...? It's unique IDs, not usernames.

@ggtylerr what do you mean by this? usernames are definitely unique-only right now.

if i open up my alt test account and try to change its username to the same as my main account, i get an error (which shows as a 409 in the console).

Reply:
I thought it was unique IDs since you ping people with them and not by their username (plus they show on profile.) If that is the case though, then I can see why.

Reply:
that's just so username changes don't suddenly break pings. discord does effectively the same thing, they're just better at hiding it in the UI.

Reply:
Added this into the document.